src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,270 @@
+        nonIntersectRect;
+
+    // calculate ratio, call callback based on new ratio value.
+    const ratio = intersectionRatio(intersectionRect, element);
+    const newThresholdSlot = getThresholdSlot(this.threshold_, ratio);
+    if (newThresholdSlot == this.prevThresholdSlot_) {
+      return null;
+    }
+    this.prevThresholdSlot_ = newThresholdSlot;
+
+    let boundingClientRect = element;
+    let rootBounds = hostViewport;
+
+    // If element inside an iframe. Every Layoutrect has already adjust their
+    // origin according to opt_iframe rect origin.
+    if (opt_iframe) {
+      // If element is inside a non-scrollable iframe. LayoutRect position is
+      // relative to iframe origin, thus relative to iframe's viewport origin,
+      // because the viewport is at the iframe origin.
+      // No need to adjust position here.
+      // To get same behavior as native IntersectionObserver set rootBounds null
+      rootBounds = null;
+    } else {
+      // If element not in an iframe.
+      // adjust all LayoutRect to hostViewport Origin.
+      intersectionRect = moveLayoutRect(intersectionRect, -hostViewport.left,
+          -hostViewport.top);
+      // The element is relative to (0, 0), while the viewport moves. So, we must
+      // adjust.
+      boundingClientRect = moveLayoutRect(boundingClientRect,
+          -hostViewport.left, -hostViewport.top);
+      // Now, move the viewport to (0, 0)
+      rootBounds = moveLayoutRect(rootBounds,
+          -hostViewport.left, -hostViewport.top);
+    }
+
+    return /** @type {!IntersectionObserverEntry} */ ({
+      time: performance.now(),
+      rootBounds: rootBounds && DomRectFromLayoutRect(rootBounds),
+      boundingClientRect: DomRectFromLayoutRect(boundingClientRect),
+      intersectionRect: DomRectFromLayoutRect(intersectionRect),
+      intersectionRatio: ratio,
+    });
+  }
+}
+
+/**
+ * Transforms a LayoutRect into a DOMRect for use in intersection observers.
+ * @param {!./layout-rect.LayoutRectDef} rect
+ * @return {!DOMRect}
+ */
+function DomRectFromLayoutRect(rect) {
+  return {
+    left: rect.left,
+    top: rect.top,
+    width: rect.width,
+    height: rect.height,
+    bottom: rect.bottom,
+    right: rect.right,
+    x: rect.left,
+    y: rect.top,
+  };
+}
+
+/**
+ * Returns the ratio of the smaller box's area to the larger box's area.
+ * @param {!./layout-rect.LayoutRectDef} smaller
+ * @param {!./layout-rect.LayoutRectDef} larger
+ * @return {number}
+ */
+function intersectionRatio(smaller, larger) {
+  return (smaller.width * smaller.height) / (larger.width * larger.height);
+}
+
+/**
+ * Returns the slot number that the current ratio fills in.
+ * @param {!Array} sortedThreshold valid sorted IoB threshold
+ * @param {number} ratio Range from [0, 1]
+ * @return {number} Range from [0, threshold.length]
+ * @visibleForTesting
+ */
+export function getThresholdSlot(sortedThreshold, ratio) {
```

Comment:
But to support testing we need to export here. I think `@visibleForTesting` will do the work for us.